### PR TITLE
Etch optimiser

### PR DESCRIPTION
**Pull request by baykaner**
- adding ml_optimiser to etch (WIP)
- refactoring vm_modules math and ml

### CODE REVIEW

File: libs/vm_modules/include/vm_modules/ml/graph.hpp [hunk 1]
```diff
@@ -28,86 +28,84 @@
...
+  void AddPlaceholder(VMPtrString const &name)
   {
-    fetch::ml::Graph<fetch::math::Tensor<float>>::AddNode<
-        fetch::ml::ops::PlaceHolder<fetch::math::Tensor<float>>>(name->str, {});
+    graph_.AddNode<fetch::ml::ops::PlaceHolder<MathTensorType>>(name->str, {});
   }
 
-  void AddFullyConnected(fetch::vm::Ptr<fetch::vm::String> const &name,
-                         fetch::vm::Ptr<fetch::vm::String> const &inputName, int in, int out)
+  void AddFullyConnected(VMPtrString const &name, VMPtrString const &inputName, int in, int out)
```
Comment (MissingNO57) [inline libs/vm_modules/include/vm_modules/ml/graph.hpp]: Naming convention: input_name

File: libs/vm_modules/include/vm_modules/ml/graph.hpp [hunk 2]
```diff
@@ -28,86 +28,84 @@
...
   {
-    fetch::ml::Graph<fetch::math::Tensor<float>>::AddNode<
-        fetch::ml::layers::FullyConnected<fetch::math::Tensor<float>>>(
+    graph_.AddNode<fetch::ml::layers::FullyConnected<MathTensorType>>(
         name->str, {inputName->str}, std::size_t(in), std::size_t(out));
   }
 
-  void AddRelu(fetch::vm::Ptr<fetch::vm::String> const &name,
-               fetch::vm::Ptr<fetch::vm::String> const &inputName)
+  void AddRelu(VMPtrString const &name, VMPtrString const &inputName)
```
Comment (MissingNO57) [inline libs/vm_modules/include/vm_modules/ml/graph.hpp]: input_name

File: libs/vm_modules/include/vm_modules/ml/optimisation/adam_optimiser.hpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      fetch::vm::VM *vm, fetch::vm::TypeId type_id,
+      fetch::vm::Ptr<fetch::vm_modules::ml::VMGraph> const &graph,
+      fetch::vm::Ptr<fetch::vm::String> const &             input_node_names,
+      fetch::vm::Ptr<fetch::vm::String> const &             output_node_names)
+  {
+    return new VMAdamOptimiser(vm, type_id, graph->graph_, {input_node_names->str},
+                               output_node_names->str);
+  }
+
+  DataType Run(fetch::vm::Ptr<fetch::vm_modules::ml::MnistDataLoader> const &loader,
```
Comment (MissingNO57) [inline libs/vm_modules/include/vm_modules/ml/optimisation/adam_optimiser.hpp:64]: Shouldn't it be for any loader?

**Review by MissingNO57 — CHANGES_REQUESTED** [review]

File: libs/vm_modules/include/vm_modules/ml/optimisation/adam_optimiser.hpp (hunk 3 above)
Comment (baykaner) [reply libs/vm_modules/include/vm_modules/ml/optimisation/adam_optimiser.hpp:64]: ideally yes; at the moment there are no other dataloader wrappers implemented in vm_modules.

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**